[WPE] Add --size / --maximized / --fullscreen parameters to MiniBrowser

nikolaszimmermann opened the pull request:
#### d72f4bd734ae263071bcd38db5ded174974a5bd4
<pre>
[WPE] Add --size / --maximized / --fullscreen parameters to MiniBrowser
<a href="https://bugs.webkit.org/show_bug.cgi?id=276072">https://bugs.webkit.org/show_bug.cgi?id=276072</a>

Reviewed by Carlos Garcia Campos.

Add --size parameter to MiniBrowser to specify the window width/height
as command line argument. This is necessary for our plans to switch WPE
perf bot testing to use MiniBrowser+new API instead of cog+old API.

Furthermore add --maximized / --fullscreen parameters supported only
within the new WPE platform API, which start MiniBrowser either
maximized or in fullscreen mode.

Doesn&apos;t affect any tests.

* Tools/MiniBrowser/wpe/main.cpp:
(parseWindowSize):
(createWebView):
(activate):
(main):

Canonical link: <a href="https://commits.webkit.org/280571@main">https://commits.webkit.org/280571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aed8fa7ba1522a778b96642b499027289a5d4912

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57005 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36333 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60625 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7448 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59133 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43956 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7638 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46168 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5235 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34120 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27028 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30900 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6534 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6453 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52859 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6805 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62306 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/918 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6909 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53426 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/921 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49258 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53467 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12599 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/779 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32162 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33247 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34332 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32993 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -58,6 +58,9 @@ static char* timeZone;
 static gboolean printVersion;
+static const char* windowSize;
+static guint windowWidth = 1280; // Default width, when using old API.
+static guint windowHeight = 720; // Default height, when using old API.
```
I would initialize these to 0, and then use the default value for old api when command line option is not used (they are still 0).

carlosgcampos replied:
They are also used in the new api when --size is not passed

nikolaszimmermann replied:
Fixed.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -82,6 +85,7 @@ static const GOptionEntry commandLineOptions[] =
 #if ENABLE_WPE_PLATFORM
     { "use-wpe-platform-api", 0, 0, G_OPTION_ARG_NONE, &useWPEPlatformAPI, "Use the WPE platform API", nullptr },
 #endif
+    { "size", 's', 0, G_OPTION_ARG_STRING, &windowSize, "Specify the window size to use, e.g. --size=\"800x600\"", nullptr },
```
For this kind of parameters that need parsing, there's `G_OPTION_ARG_CALLBACK`, so that a callback is emitted with the string and you can parse it and initialize the global values

nikolaszimmermann replied:
Nice, will switch to that.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -514,6 +521,34 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
+
+    gchar** windowSizes = g_strsplit(windowSize, "x", 2);
+    if (!windowSizes || !windowSizes[0] || !windowSizes[1]) {
+        g_warning("Failed to parse --size command line argument - use \"<size>x<height>\" without any additional whitespace.");
```
With the option callback you have a `GError` to set instead of `g_warning`

nikolaszimmermann replied:
Done.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -514,6 +521,34 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
+    if (!windowSize)
+        return FALSE;
+
+    gchar** windowSizes = g_strsplit(windowSize, "x", 2);
```
`windowSizes` needs to be freed with `g_strfreev()`

nikolaszimmermann replied:
Oh good catch, I used `GUniquePtr<const char*>` before which is not available in `Tools/MiniBrowser`. Will fix.

nikolaszimmermann replied:
Will switch to `g_auto(GStrv) = ...` as suggested by @GeorgesStavracas :-)

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -568,7 +603,16 @@ int main(int argc, char *argv[])
+        if (useWPEPlatformAPI) {
+            windowWidth = 1024;
+            windowHeight = 768;
+        }
```
I understand now the comment on the initialization. I still think it's better to initialize to 0 and only resize if they are set.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -514,6 +521,34 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
+        return FALSE;
+    }
+
+    windowWidth = static_cast<guint>(parsedWidth);
+    windowHeight = static_cast<guint>(parsedHeight);
```
We should probably check that both are greater than 0.

nikolaszimmermann replied:
Indeed, will adapt.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -276,8 +315,11 @@ static WebKitWebView* createWebView(WebKitWebView* webView, WebKitNavigationActi
     if (auto* wpeView = webkit_web_view_get_wpe_view(newWebView)) {
+        auto* wpeToplevel = wpe_view_get_toplevel(wpeView);
+        if (windowWidth > 0 && windowHeight > 0)
+            wpe_toplevel_resize(wpeToplevel, windowWidth, windowHeight);
```
I don't think we should apply the size in this case, we are expected to use `WebKitWindowProperties` in this case, connecting to ready-to-show signal on the new web view. I see it's what we currently do, so it's probably ok to do this for the new API too, and handle the window properties in a follow up patch.

nikolaszimmermann replied:
As discussed, makes no sense to use the specified --size parameters for secondary windows that are opened.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -474,8 +516,15 @@ static void activate(GApplication* application, WPEToolingBackends::ViewBackend*
+        auto* wpeToplevel = wpe_view_get_toplevel(wpeView);
+        if (windowWidth > 0 && windowHeight > 0)
+            wpe_toplevel_resize(wpeToplevel, windowWidth, windowHeight);
+        if (windowMaximized)
+            wpe_toplevel_maximize(wpeToplevel);
+        if (windowFullscreen)
+            wpe_toplevel_fullscreen(wpeToplevel);
```
fullscreen and maximizez should be mutually exclusive.

nikolaszimmermann replied:
Fixed.

```
+    if (windowMaximized && windowFullscreen) {
+        g_printerr("You cannot specify both --maximized and --fullscreen, these options are mutually exclusive.");
+        return 1;
+    }
```

nikolaszimmermann committed:
[WPE] Add --size / --maximized / --fullscreen parameters to MiniBrowser
https://bugs.webkit.org/show_bug.cgi?id=276072

Reviewed by Carlos Garcia Campos.

Add --size parameter to MiniBrowser to specify the window width/height
as command line argument. This is necessary for our plans to switch WPE
perf bot testing to use MiniBrowser+new API instead of cog+old API.

Furthermore add --maximized / --fullscreen parameters supported only
within the new WPE platform API, which start MiniBrowser either
maximized or in fullscreen mode.

Doesn't affect any tests.

* Tools/MiniBrowser/wpe/main.cpp:
(parseWindowSize):
(createWebView):
(activate):
(main):

Canonical link: https://commits.webkit.org/280571@main
